File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    int visited[] = new int[10];
    Set<Integer> set = new HashSet<>();
    
    //소수 판별 
    public boolean isPrime(int n){
        if(n==1 || n==0) return false;
        for(int i=2; i<n; i++) if(n%i==0) 
            return false;
        return true;
    }
    
    //DFS로 numbers로 만들 수 있는 모든 경우의 수 찾음 -> Set으로 중복 제거 
    public void DFS(String numbers, int cnt, String result){
        if(cnt == numbers.length()){
            if(result != ""){
                set.add(Integer.parseInt(result));
            }
            return;
        }
        
        for(int i=0; i<numbers.length(); i++){
            if(visited[i]==0){
                visited[i] = 1;
                DFS(numbers, cnt+1, result+numbers.charAt(i));
                visited[i] = 0;
                DFS(numbers, cnt+1, result);
            }
        }
    }
    
    //결과 도출
    public int solution(String numbers) {
        int answer = 0;
        //DFS로 모든 경우의 수 찾고 Set에 값 정리 
        DFS(numbers, 0, "");
        for(int s : set){
            //Set 돌면서 해당 값이 소수인지 판별 
            if(isPrime(s)) answer++;
        }
        return answer;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

vector<int> answerList;
bool isPrime(int n){
    if(n==0 || n==1) return false;
    for(int i=2; i<n; i++){
        if(n%i==0) return false;
    }
    return true;
}
void dfs(string res, vector<char> nums, vector<bool> visited){
    if(res.length()>nums.size()) return;
    else{
        //cout<<"res:"<<res<<endl;
        if(res.length()>=1 && isPrime(stoi(res))) answerList.push_back(stoi(res));
        for(int i=0; i<nums.size(); i++){
            if(!visited[i]){
                visited[i] = true;
                dfs(res+nums[i],nums,visited);
                visited[i] = false;
            }
        }
    }
}
int solution(string numbers) {
    vector<char> nums;
    for(char c : numbers) nums.push_back(c);
    vector<bool> visited(numbers.length());
    dfs("",nums,visited);
    
    sort(answerList.begin(), answerList.end());
    auto it = unique(answerList.begin(), answerList.end());
    answerList.erase(it, answerList.end());
    return answerList.size();
}